components/style/rule_tree/mod.rs

```diff
@@ -0,0 +1,491 @@
+            }
+        }
+        write!(writer, "\n");
+        for child in self.children.iter() {
+            child.get().dump(writer, indent + INDENT_INCREMENT);
+        }
+    }
+}
+
+#[derive(Clone)]
+struct WeakRuleNode {
+    ptr: *mut RuleNode,
+}
+
+pub struct StrongRuleNode {
+    ptr: *mut RuleNode,
+}
+
+#[cfg(feature = "servo")]
+impl HeapSizeOf for StrongRuleNode {
+    fn heap_size_of_children(&self) -> usize { 0 }
+}
+
+
+impl StrongRuleNode {
+    fn new(n: Box<RuleNode>) -> Self {
+        debug_assert!(n.parent.is_none() == n.rule.is_none());
+
+        let ptr = Box::into_raw(n);
+
+        debug!("Creating rule node: 0x{:x}", ptr as usize);
+
+        StrongRuleNode {
+            ptr: ptr,
+        }
+    }
+
+    fn downgrade(&self) -> WeakRuleNode {
+        WeakRuleNode {
+            ptr: self.ptr,
+        }
+    }
+
+    fn next(&self) -> Option<WeakRuleNode> {
+        // FIXME(emilio): Investigate what ordering can we achieve without
+        // messing things up.
+        let ptr = self.get().next.load(Ordering::SeqCst);
+        if ptr.is_null() {
+            None
+        } else {
+            Some(WeakRuleNode {
+                ptr: ptr
+            })
+        }
+    }
+
+    fn parent(&self) -> Option<StrongRuleNode> {
+        self.get().parent.clone()
+    }
+
+    fn ensure_child(&self,
+                    root: &StrongRuleNode,
+                    rule: &Arc<StyleRule>,
+                    importance: Importance) -> StrongRuleNode {
+        let mut last = None;
+        let mut iter = self.get().children.iter();
+        while let Some(current) = iter.next() {
+            if current.get().importance == importance &&
+               arc_ptr_eq(current.get().rule.as_ref().unwrap(), rule) {
+                return current;
+            }
+            last = Some(current);
+        }
+
+        let node = Box::new(RuleNode::new(root.clone(),
+                                          self.clone(),
+                                          rule.clone(),
+                                          importance));
+        let new_ptr = &*node as *const _ as *mut RuleNode;
+
+        loop {
+            let strong;
+
+            {
+                let next_ptr = match last {
+                    Some(ref l) => &l.get().next,
+                    None => &self.get().children.head,
+                };
+
+                let existing =
+                    next_ptr.compare_and_swap(ptr::null_mut(),
+                                              new_ptr,
+                                              Ordering::SeqCst);
+
+                if existing == ptr::null_mut() {
+                    // Now we know we're in the correct position in the child list,
+                    // we can set the back pointer, knowing that this will only be
+                    // accessed again in a single-threaded manner when we're
+                    // sweeping possibly dead nodes.
```

Comment:
Can you add a comment to `prev` specifying this very special memory constraint? This is *tricky*.